### ISSUE TITLE
Bug calculating cyclomatic complexity

### ISSUE THREAD

**Issue by msiemens** [Manishearth/rust-clippy#491]
Clippy asked me to file a bug report. It fails with this message:

```plain
src\util\io.rs:10:1: 36:2 note: Clippy encountered a bug calculating cyclomatic complexity (hide this message with `#[allow(cyclomatic_complexity)]`): cc = 1, arms = 2. Please file a bug report.
src\util\io.rs:10 pub fn read_file(input_path: &str) -> String {
src\util\io.rs:11     let mut file = match File::open(&Path::new(input_path)) {
src\util\io.rs:12         Ok(f) => f,
src\util\io.rs:13         Err(err) => {
src\util\io.rs:14             fatal!("Can't open {}: {}", input_path, err);
src\util\io.rs:15             session().abort();
                  ...
```

The file mentioned in the snippet:

```rust
/// Read a file and return it's contents
//! File I/O related helpers

use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;
use driver::session;


/// Read a file and return it's contents
pub fn read_file(input_path: &str) -> String {
    let mut file = match File::open(&Path::new(input_path)) {
        Ok(f) => f,
        Err(err) => {
            fatal!("Can't open {}: {}", input_path, err);
            session().abort();
        }
    };

    let mut bytes = Vec::new();

    match file.read_to_end(&mut bytes) {
        Ok(..) => {},
        Err(_) => {
            fatal!("Can't read {}", input_path);
            session().abort();
        }
    };

    match String::from_utf8(bytes) {
        Ok(contents) => return contents,
        Err(_) => {
            fatal!("{} is not UTF-8 encoed", input_path);
            session().abort();
        }
    }
}
```

Versions:

```
rustc 1.7.0-nightly (81ae8be71 2015-12-09)
cargo 0.7.0-nightly (1af03be 2015-12-08)

[[package]]
name = "clippy"
version = "0.0.30"
source = "git+https://github.com/Manishearth/rust-clippy#3e7677ff319a3b2989c29bc57d258e7b8f39fed6"
```

**Comment by Manishearth** [comment]
cc @oli-obk 

Is this the same SEME error?

**Comment by mrmonday** [comment]
I'm encountering the same issue, I don't have time to reduce a test case though. To reproduce:

```
git clone https://github.com/libpnet/libpnet.git
cd libpnet
cargo test --no-run
```

Error:
```
src/test.rs:292:1: 393:2 note: Clippy encountered a bug calculating cyclomatic complexity (hide this message with `#[allow(cyclomatic_complexity)]`): cc = 1, arms = 2. Please file a bug report.
src/test.rs:292 fn layer2() {
src/test.rs:293     use datalink::{DataLinkChannelType, datalink_channel};
src/test.rs:294     use packet::ethernet::{EtherTypes, EthernetPacket, MutableEthernetPacket};
src/test.rs:295 
src/test.rs:296     const MIN_PACKET_SIZE: usize = 64;
src/test.rs:297     const ETHERNET_HEADER_LEN: usize = 14;
                ...
```

Versions:
```
rustc 1.7.0-nightly (81ae8be71 2015-12-09)
cargo 0.7.0-nightly (1af03be 2015-12-08)
clippy 0.0.30
Linux 4.2.5-1-ARCH x86_64
```

**Comment by oli-obk** [comment]
looking into it

**Comment by llogiq** [comment]
I think this is.the first time clippy diagnosed itself. :smile:

**Comment by oli-obk** [comment]
does anyone know how to figure out if an expression diverges?

**Comment by llogiq** [comment]
This is called the halting problem and is intractable in the general case. However, macro expansions with names like `panic` are a strong hint that something diverges, then there are diverging return types...

**Comment by oli-obk** [comment]
I'm fine with knowing there's no non-diverging path. I mean, currently the `unreachable_code` lint can already prove that an expression diverges guaranteed.

... oh while writing this I noticed I should simply check what the `unreachable_code` lint does

**Comment by Manishearth** [comment]
I think there's a flag on `Ty`.

It's not exactly the halting problem since you're just checking if rustc thinks that the expr diverges, which means "it is possible to prove that this diverges with the info available to rustc today", which is not the halting problem.

**Comment by Manishearth** [comment]
`type_var_diverges` is what we want but we don't have an inference/fulfilment context.
